### ISSUE TITLE
Incompatible signatures incorrectly reported in the case of generic method unions

### ISSUE THREAD

**Issue by craigphicks** [microsoft/TypeScript#57639]
### 🔎 Search Terms

This expression is not callable.
  Each member of the union type '{ (cb: (x: number) => number): number[]; <U>(cb: (x: number) => U): U[]; } | { (cb: (x: string) => string): string[]; <U>(cb: (x: string) => U): U[]; }' has signatures, but none of those signatures are compatible with each other.ts(2349)

### 🕗 Version & Regression Information

- This is the behavior in every version I tried

### ⏯ Playground Link

https://www.typescriptlang.org/play?#code/HYQwtgpgzgDiDGEAEwoAYkG8BQ2CWwALhAE4BmCyAKtIQIx1oA8VAfFtkl0mQBTwAjAFy8AHkKoBKALyspEgNoBdANyduZJgFVW-YWIkzWWyUK3K1AX1wATCPAA2IEsngB7VISTOSQpDSh6RiZgAFcwAVJ2AB9-WgZmQJICAHNWNXdPJBcoUIcvaW8SEgA6PlEkWSRRSRUkAHp6pFISNxJsRv8ACzwoZtEYHKg8DyRelDcveBAHJwEHCBKOpqQAUQQupEgI0iQ3MiRCLuRQ4BHgQ4BPGGQAckwkPT8DFHDIkklK9jCdj78f94WJDaXSCZ7iV6-T5VEx+cyqJCWJCxB5PR4QpKpaHsTHAFKmJC4lJAkFol5E7FIWFUoGWW5ILogPrDFKgQihHIAGiQAlCXmAHmQ+0OXTcUGQLLZHOgRVcbjAcEIeHmyAA7ngjs0NnsjqQSrwAEwAZgALABOSS4bDWXCgSCwSgoKB0Dj4IikCiIOKBBh0FjsHDcHhPXggCTc8RSWTyKgWdRcTQ6ENhrQRwyyVMEDVmUzwqy2exOFxITKBIq+b1BP0AqLIyu+pgCPApAiEdLYUteHJ5Arlsq8UNCJstogRykgJAAKiQACFmwBJIhiSTcujAWoNJotNrLbrjCADIbnMZ9AVTGZzBZLTpcdbwTbbd57A66pCnY+Ea53VFgx5hyHvGmAGkJSNZ-MBJAkkmv6DlSQFgZSqZjGchBwgSeaInWP7CH+fjDq2QH4UQlJEYQBKkVBoI4bBSEQqRiHclmqFUuhtL0oyzLNlKXI8nyEzAEKL6iuKhJcSA7I5LKJbyoqyoLEg6qahA2qTMcpSGqaFrWtgQA

### 💻 Code

```ts
namespace ns0 {

interface Test110<T> {
    f(cb:(x:T)=>T):T[];
    f<U>(cb:(x:T)=>U):U[];
}

declare const arr: Test110<number> | Test110<string>;
const result = arr.f(x => x); // error
// This expression is not callable.
//  Each member of the union type '{ (cb: (x: number) => number): number[]; <U>(cb: (x: number) => U): U[]; } | { (cb: (x: string) => string): string[]; <U>(cb: (x: string) => U): U[]; }' has signatures, but none of those signatures are compatible with each other.(2349)


}


namespace ns1 {

interface Test111<T> {
    f(cb:(a:T, x:T)=>T):T[];
    f<U>(cb:(a:U, x:T)=>U,init:U):U[];
}

declare const arr: Test111<number> | Test111<bigint>;
const result = arr.f((a:bigint, x) => a * BigInt(x), 1n); // error
// This expression is not callable.
//   Each member of the union type '{ (cb: (a: number, x: number) => number): number[]; <U>(cb: (a: U, x: number) => U, init: U): U[]; } | { (cb: (a: bigint, x: bigint) => bigint): bigint[]; <U>(cb: (a: U, x: bigint) => U, init: U): U[]; }' has signatures, but none of those signatures are compatible with each other.(2349)
}
```

### 🙁 Actual behavior

Errors as shown

### 🙂 Expected behavior

No errors

### Additional information about the issue

Fix available at pull #57636

**Comment by RyanCavanaugh** [comment]
The proposed merging being implied here isn't safe because we can't say for certain which overload gets picked in each branch of the union without doing overload resolution on each constituent of the union, which is a combinatorially-explosive operation in the worst case.

**Comment by craigphicks** [comment]
@RyanCavanaugh - It is possible to test the general case first, and only if that fails do the more complete analysis.  In fact that exact algorithm is already used in some cases.  For example the `push` case above is testing every overload after testing the general first, which fails.  So there is nothing new in doing that.

**Comment by fatcerberus** [comment]
@craigphicks He means testing every overload *for each member of a union*, i.e. the kind of logic that would be required for this to work:

```ts
declare function foo(x: string): void;
declare function foo(x: number): void;
declare const x: string | number;
foo(x);  // <-- this is an error today: no overload matches
```

...which is indeed combinatorially explosive in the number of parameters and the size of the union(s) involved.

**Comment by craigphicks** [comment]
@fatcerebus - I think the problem Ryan is specifying is that `T` and `U` in the following can have a large number of union members, and that would be multiplied by the number of overloads.  

```
interface Test111<T> {
    f(cb:(a:T, x:T)=>T):T[];
    f<U>(cb:(a:U, x:T)=>U,init:U):U[];
}
```

So the total worst case number is ||T|| x ||U|| x ||f||.  While ||f|| is not going to grow unexpectedly, the size of ||T|| and ||U|| are not something controlled by the original function writer.  Of course that problem with the size ||T|| x ||U|| can occur even with when ||f||===1.  So it doesn't make sense to be concerned only about the size of ||f||, and not worry about the size of ||T|| x  ||U||.

Setting a total  limit on ||U|| x ||T|| x ||f|| could make sense, and that would include the case of ||f||===1.

Re: The general case I mentioned:
```
declare arr: string[]|number[]
declare x: string|number;
arr.map(x=>x)
```
The algorithm first computes the general case 
```
arr.map(x=>x)
```
substituting `string|number->x`, which passes, and so finishes the computation without iterating over the union elements.

To be exact there are 3 possible signatures potentially reviewed:
string|number -> string|number : general case
number -> number                    : element 0 only
string -> string                           : element 1 only

Only the first is "checked" in the shown case.

That is a common case - the mode case - and the average cost matters.

**Comment by fatcerberus** [comment]
Editing performance definitely matters more than you’d think - see: frequent complaints from the TS maintainers themselves about slowdown and poor responsiveness while editing ‘checker.ts`.

**Comment by RyanCavanaugh** [comment]
If you merge 16,000 PRs that each incur a 0.01% perf penalty (i.e. smaller than you can measure), your code is now 5x slower than before and people start yelling at you to rewrite the project in Rust. Ask me how I know 🥲

**Comment by craigphicks** [comment]
@fatcerberus @RyanCavanaugh - Out of line response to my comment pointing out that processing of actual errors (which including building strings from types etc. for the error message) is a miniscule burden overall compared the 99.99% which is type checking of non-errors.

**Comment by craigphicks** [comment]
@RyanCavanaugh 
> If you merge 16,000 PRs that each incur a 0.01% perf penalty (i.e. smaller than you can measure), your code is now 5x slower than before and people start yelling at you to rewrite the project in Rust. Ask me how I know 🥲

I understand.  Since it all server side, the most expensive computation patterns could be modularized and rewritten in a non-interpretive language.

**Comment by fatcerberus** [comment]
@craigphicks “Rewrite it in Rust” is kind of a meme - whenever a program gets too slow, the stereotype is that all the Rust nerds will come out of the woodwork and start clamoring that the only way to fix it is to rewrite the whole thing in Rust.  Ryan wasn’t genuinely proposing that anything be written in Rust. :wink:

**Comment by craigphicks** [comment]
Interesting reading:
- [How to Use Rust with Node.js When Performance Matters](https://blog.risingstack.com/how-to-use-rust-with-node-when-performance-matters/)
    - Covers a toy program running Fibonacci computation, calling a rust submodule from nodejs.  Reports about a 2/3 speedup.
- [My JavaScript is Faster than Your Rust](https://medium.com/@jbyj/my-javascript-is-faster-than-your-rust-5f98fe5db1bf)
    -  The title only described the initial Rust implementation the author developed.  After a lot of work on memory management, the author made an optimized Rust implementation with about a 2/3 speedup over pure js. The author also points out the ease of maintenance when using pure js and no Rust.  

**Premature optimization is the root of all evil (or at least most of it) in programming** (Hoare, Knuth).  Which means, if TypeScript were to be optimized with Rust, it would only make sense if the design had more or less converged, and all bugs fixed.

My understanding is that the V8 JS engine uses up to multiple threads to handle the GC, even though the program (except for GC) is being run on a single thread.  So that could be a high bar to match if the memory management were to be be implemented in Rust.   Or maybe there is someway to let memory management be handled the js level, and just call Rust for computing on already allocated memory.  

I'm not advocating such an optimization here - just think the topic is interesting.
